src/ossl/aes.rs

```diff
@@ -1660,3 +1658,232 @@ impl Decryption for AesOperation {
+    in_use: bool,
+    _key: AesKey,
+    _mac: EvpMac,
+    ctx: EvpMacCtx,
+    maclen: usize,
+}
+
+impl AesCmacOperation {
+    fn register_mechanisms(mechs: &mut Mechanisms) {
+        for ckm in &[CKM_AES_CMAC, CKM_AES_CMAC_GENERAL] {
+            mechs.add_mechanism(*ckm, new_mechanism(CKF_SIGN | CKF_VERIFY));
+        }
+    }
+
+    fn init(mech: &CK_MECHANISM, key: &Object) -> KResult<AesCmacOperation> {
+        let maclen = match mech.mechanism {
+            CKM_AES_CMAC_GENERAL => {
+                if mech.ulParameterLen as usize
+                    != ::std::mem::size_of::<CK_MAC_GENERAL_PARAMS>()
+                {
+                    return err_rv!(CKR_ARGUMENTS_BAD);
+                }
+                let val: usize =
+                    unsafe { *(mech.pParameter as CK_MAC_GENERAL_PARAMS_PTR) }
+                        as usize;
+                if val > AES_BLOCK_SIZE {
+                    return err_rv!(CKR_MECHANISM_PARAM_INVALID);
+                }
+                val
+            }
+            CKM_AES_CMAC => {
+                if mech.ulParameterLen != 0 {
+                    return err_rv!(CKR_ARGUMENTS_BAD);
+                }
+                AES_BLOCK_SIZE
+            }
+            _ => return err_rv!(CKR_MECHANISM_INVALID),
+        };
+        let mackey = object_to_raw_key(key)?;
+        let mut mac = match EvpMac::from_ptr(unsafe {
+            EVP_MAC_fetch(
+                get_libctx(),
+                name_as_char(OSSL_MAC_NAME_CMAC),
+                std::ptr::null(),
+            )
+        }) {
+            Ok(em) => em,
+            Err(_) => return err_rv!(CKR_DEVICE_ERROR),
+        };
+        let mut ctx = match EvpMacCtx::from_ptr(unsafe {
+            EVP_MAC_CTX_new(mac.as_mut_ptr())
+        }) {
+            Ok(emc) => emc,
+            Err(_) => return err_rv!(CKR_DEVICE_ERROR),
+        };
+        let params = OsslParam::new()
+            .add_const_c_string(
+                name_as_char(OSSL_MAC_PARAM_CIPHER),
+                match mackey.raw.len() {
+                    16 => name_as_char(CIPHER_NAME_AES128),
+                    24 => name_as_char(CIPHER_NAME_AES192),
+                    32 => name_as_char(CIPHER_NAME_AES256),
+                    _ => return err_rv!(CKR_KEY_INDIGESTIBLE),
+                },
+            )?
+            .finalize();
+
+        if unsafe {
+            EVP_MAC_init(
+                ctx.as_mut_ptr(),
+                mackey.raw.as_ptr(),
+                mackey.raw.len(),
+                params.as_ptr(),
+            )
+        } != 1
+        {
+            return err_rv!(CKR_DEVICE_ERROR);
+        }
+        Ok(AesCmacOperation {
+            finalized: false,
+            in_use: false,
+            _key: mackey,
+            _mac: mac,
+            ctx: ctx,
+            maclen: maclen,
+        })
+    }
+
+    fn begin(&mut self) -> KResult<()> {
+        if self.in_use {
+            return err_rv!(CKR_OPERATION_NOT_INITIALIZED);
```

Comment:
Is this appropriate error? Should it by `CKR_OPERATION_ACTIVE` or something similar?

Reply:
I've used CKR_OPERATION_NOT_INITIALIZED everywhere, not just in this code.
I guess we could use CKR_OPERATION_ACTIVE, but need to check the spec for the exact semantics of when it is ok to return it.
Keep in mind this is incorrect use, so 99.9% of applications should enver see this error, I would not hand this PR on this, but perhaps file an issue to review all use of CKR_OPERATION_* errors?